### PR TITLE
fix(simple-list): update state properly when external props change

### LINKED ISSUE #2316: [SimpleList] Cannot change Icon component without throwing away and regenning the list
### What package is this for?

- [x] React
- [ ] Angular

### Describe the bug

Because the simple list keeps the items property in local state it doesn't update if the items list changes.  Specifically I change the icon based on an interaction with the list, but the list never updates.  I am forced to change the key of the list so React throws it away

### To Reproduce

Pass the simple list a set of items with icons.  Now change the icons based on some interaction with the page (maybe a button at the top).  The icons will never change in the simple list

### PR DESCRIPTION
**Pull request by kevinsperrine**
Closes #2316 

**Summary**

Previously, the whole list needed to be thrown out when any props changed, but now I've added useEffects to update the state as new items are passed or as pageSize change, for example. It should maintain the previous state, so if you were on Page 2 and new items are added you'll stay on Page 2, but if items are removed you'll reset back to page 1. If you previously had a search value and you add more items the search value will be maintained.

**Change List (commits, features, bugs, etc)**

- added useEffects to update state when items or pagination props change
- added new tests to cover all this functionality and increase coverage even more

**Acceptance Test (how to verify the PR)**

- Open the Basic SimpleList story
- change the number of items rendering to 10
- confirm only 10 items shown
- change pageSize to sm 
- confirm only 5 items shown
- move to page 2
- change the number of items rendering to 20
- confirm still on page 2
- move to page 3
- confirm 20 items are being rendered
- Repeat similar steps with search and confirm items shown and pagination is maintained when changing other external props

**Regression Test (how to make sure this PR doesn't break old functionality)**

- All previous functionality works as expected.

**Things to look for during review**

- [ ] Make sure all references to `iot` or `bx` class prefix is using the prefix variable
- [ ] (React) All major areas have a `data-testid` attribute. New test ids should have test written to ensure they are not changed or removed.
- [ ] UI should be checked in RTL mode to ensure the proper handling of layout and text.
- [ ] All strings should be translatable.
- [ ] The code should pass a11y scans (The storybook a11y knob should show no violations). New components should have a11y test written.
- [ ] Unit test should be written and should have a coverage of 90% or higher in all areas.
- [ ] All components should be passing visual regression test. For new styles or components either a visual regression test should be written for all permutations or the base image updated.
- [ ] Changes or new components should either write new or update existing documentation.
- [ ] PR should link and close out an existing issue

### CODE REVIEW

File: packages/react/src/components/List/SimpleList/SimpleList.jsx [hunk 1]
```diff
@@ -102,21 +102,97 @@ const SimpleList = ({
...
+      setCurrentPageNumber(page);
+      setItemsToShow(currentItemsOnPage);
+    },
+    [filteredItems, rowPerPage]
+  );
+
+  const maxPage = Math.ceil(numberOfItems / rowPerPage);
+
+  /**
+   * Then the items array updates, make sure to update the filteredItems, but also re-apply
```
Comment (bjornalm) [inline packages/react/src/components/List/SimpleList/SimpleList.jsx:167]: nit: spelling?

File: packages/react/src/components/List/SimpleList/SimpleList.test.jsx [hunk 2]
```diff
@@ -266,4 +267,171 @@ describe('SimpleList', () => {
...
+    expect(screen.getByText('Item 5')).toBeVisible();
+    expect(screen.getByText('Item 15')).toBeVisible();
+    expect(screen.getByText('Item 25')).toBeVisible();
+    expect(screen.getByText('Item 35')).toBeVisible();
+    expect(screen.getByText('Item 45')).toBeVisible();
+    expect(screen.getByText('Page 1')).toBeVisible();
+    userEvent.click(screen.getByRole('button', { name: 'Next page' }));
+    expect(screen.getByText('Item 50')).toBeVisible();
+    expect(screen.getByText('Page 2')).toBeVisible();
+    expect(screen.getByRole('button', { name: 'Next page' })).toHaveClass(
```
Comment (bjornalm) [inline packages/react/src/components/List/SimpleList/SimpleList.test.jsx:354]: nit: I guess .toBeDisabled() isn't working on these buttons?

**Review by bjornalm — APPROVED** [review]
Nice job, just questions and minor things

File: packages/react/src/components/List/SimpleList/SimpleList.jsx [hunk 3]
```diff
@@ -102,21 +102,97 @@ const SimpleList = ({
...
+
+  const handleSearch = useCallback(
+    (searchTerm) => {
+      setSearchValue(searchTerm);
+      const searchFilteredItems = items.filter((item) => {
+        if (item.content.value !== '' && item.content.value !== undefined) {
+          if (
+            item.content.secondaryValue !== '' &&
+            item.content.secondaryValue !== undefined &&
+            typeof item.content.secondaryValue === 'string'
```
[the remark this answers is not in the document]
Comment (kevinsperrine) [reply packages/react/src/components/List/SimpleList/SimpleList.jsx:134]: yeah, I can remove it. It was just copy/pasted from the inlined onChange function below.

**Review by davidicus — APPROVED** [review]

**Pull request merged**